File: packages/python/procedurals/occupied_lattice_paths.py
"""Retained shared-occupancy cardinal lattice paths independently specified by the
shared path.occupied-lattice-paths-2d contract.

Motivating note: survey/out/2019/generativos/tata/notes.md. This deliberately does
not reproduce tata's omitted starts, random proposals, scene stream, or rendering;
no artistic dimensions/starts/step range is established, see the catalog contract
for evidence.
"""
from array import array
import math

__all__ = ["LatticeError", "occupied_lattice_paths_2d"]
_MAX_CELLS = 1073741823
_MAX_STEPS = 1073741822
_MAX_SAFE = 9007199254740991
_UINT32_MAX = 4294967295
_MASK32 = (1 << 32) - 1
_MASK64 = (1 << 64) - 1


class LatticeError(ValueError):
    """Stable catalog code: INVALID_INPUT, WORK_LIMIT_EXCEEDED, INVALID_INDEX,
    INDEX_OUT_OF_RANGE or INVALID_OUTPUT."""
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def _fail(code):
    raise LatticeError(code)


def _integral(value):
    if type(value) not in (int, float) or type(value) is bool:
        _fail("INVALID_INPUT")
    if type(value) is float and (not math.isfinite(value) or not value.is_integer()):
        _fail("INVALID_INPUT")
    return value


def _bounded_int(value, lo, hi):
    n = _integral(value)
    if n < lo or n > hi:
        _fail("INVALID_INPUT")
    return int(n)


def _uint32(value):
    return _bounded_int(value, 0, _UINT32_MAX)


def _pair(value):
    if type(value) is not list or len(value) != 2:
        _fail("INVALID_INPUT")
    return value


def _record_either_key(value):
    if type(value) is not dict or len(value) != 1:
        _fail("INVALID_INPUT")
    return value


def _read_random(value):
    record = _record_either_key(value)
    has_seed, has_state = "seed" in record, "state" in record
    if has_seed == has_state:
        _fail("INVALID_INPUT")
    if has_seed:
        return _uint32(record["seed"]), None
    words = record["state"]
    if type(words) is not list or len(words) != 4:
        _fail("INVALID_INPUT")
    state = [_uint32(word) for word in words]
    if state == [0, 0, 0, 0]:
        _fail("INVALID_INPUT")
    return 0, state


class _Xoshiro:
    """Private xoshiro128**1.1 stream with two SplitMix64 expansion words; same
    algorithm as procedurals.quadrant_partition._Stream, extended to accept an
    explicit caller-supplied [s0,s1,s2,s3] state that bypasses expansion."""
    def __init__(self, seed, state):
        if state is not None:
            self._s = list(state)
            return
        x = seed + 0x9E3779B97F4A7C15
        a = self._split(x)
        b = self._split(x + 0x9E3779B97F4A7C15)
        self._s = [a & _MASK32, a >> 32, b & _MASK32, b >> 32]

    @staticmethod
    def _split(x):
        x &= _MASK64
        x = ((x ^ (x >> 30)) * 0xBF58476D1CE4E5B9) & _MASK64
        x = ((x ^ (x >> 27)) * 0x94D049BB133111EB) & _MASK64
        return (x ^ (x >> 31)) & _MASK64

    def next_u32(self):
        s = self._s
        value = (s[1] * 5) & _MASK32
        result = ((((value << 7) | (value >> 25)) & _MASK32) * 9) & _MASK32
        t = (s[1] << 9) & _MASK32
        s[2] ^= s[0]
        s[3] ^= s[1]
        s[1] ^= s[2]
        s[0] ^= s[3]
        s[2] ^= t
        s[3] = ((s[3] << 11) | (s[3] >> 21)) & _MASK32
        return result

    def choose(self, count):
        """floor(u32 * count / 2^32); consumes exactly one word for count in 1..4."""
        return (self.next_u32() * count) >> 32

    def state(self):
        return self._s.copy()


class _LatticePaths:
    __slots__ = ("_x", "_y", "_offsets", "_reasons", "_random_state")

    def __init__(self, x, y, offsets, reasons, random_state):
        self._x = x
        self._y = y
        self._offsets = offsets
        self._reasons = reasons
        self._random_state = random_state

    @property
    def path_count(self):
        return len(self._reasons)

    def _path_index(self, path):
        if type(path) not in (int, float) or type(path) is bool:
            _fail("INVALID_INDEX")
        if type(path) is float and (not math.isfinite(path) or not path.is_integer()):
            _fail("INVALID_INDEX")
        if path < 0 or path > _MAX_SAFE:
            _fail("INVALID_INDEX")
        path = int(path)
        if path >= len(self._reasons):
            _fail("INDEX_OUT_OF_RANGE")
        return path

    def _cell_index(self, path, cell):
        if type(cell) not in (int, float) or type(cell) is bool:
            _fail("INVALID_INDEX")
        if type(cell) is float and (not math.isfinite(cell) or not cell.is_integer()):
            _fail("INVALID_INDEX")
        if cell < 0 or cell > _MAX_SAFE:
            _fail("INVALID_INDEX")
        cell = int(cell)
        length = self._offsets[path + 1] - self._offsets[path]
        if cell >= length:
            _fail("INDEX_OUT_OF_RANGE")
        return cell

    def path_length_at(self, path):
        p = self._path_index(path)
        return self._offsets[p + 1] - self._offsets[p]

    def cell_at(self, path, cell):
        """Return a fresh detached [x, y] pair."""
        p = self._path_index(path)
        c = self._cell_index(p, cell)
        at = self._offsets[p] + c
        return [self._x[at], self._y[at]]

    def cell_into(self, path, cell, output, offset=0):
        """Validate path, cell and destination before writing; return output."""
        p = self._path_index(path)
        c = self._cell_index(p, cell)
        if not (type(output) is list or (type(output) is array and output.typecode == "i" and output.itemsize == 4)):
            _fail("INVALID_OUTPUT")
        if type(offset) not in (int, float) or type(offset) is bool:
            _fail("INVALID_OUTPUT")
        if type(offset) is float and (not math.isfinite(offset) or not offset.is_integer()):
            _fail("INVALID_OUTPUT")
        offset = int(offset)
        if offset < 0 or offset > len(output) - 2:
            _fail("INVALID_OUTPUT")
        at = self._offsets[p] + c
        output[offset] = self._x[at]
        output[offset + 1] = self._y[at]
        return output

    def completion_reason_at(self, path):
        return self._reasons[self._path_index(path)]

    def random_state(self):
        """Return a fresh detached uint32 [s0,s1,s2,s3] state."""
        return self._random_state.copy()

    def to_values(self):
        """Export detached paths, completion reasons and post-call state."""
        paths = []
        for p in range(len(self._reasons)):
            path = []
            for i in range(self._offsets[p], self._offsets[p + 1]):
                path.append([self._x[i], self._y[i]])
            paths.append(path)
        return {"paths": paths, "completionReasons": list(self._reasons), "randomState": self._random_state.copy()}


def occupied_lattice_paths_2d(config):
    """Generate ordered retained cardinal cell paths whose cells are claimed by
    one call-local occupancy set.

    Implements path.occupied-lattice-paths-2d 0.1.0 independently of source code.
    """
    if type(config) is not dict or set(config) != {"dimensions", "starts", "maxSteps", "maxCells", "random"}:
        _fail("INVALID_INPUT")
    dimensions = _pair(config["dimensions"])
    columns = _bounded_int(dimensions[0], 1, 2147483647)
    rows = _bounded_int(dimensions[1], 1, 2147483647)
    starts = config["starts"]
    if type(starts) is not list:
        _fail("INVALID_INPUT")
    if len(starts) > _MAX_CELLS:
        _fail("INVALID_INPUT")
    for item in starts:
        start = _pair(item)
        sx = _integral(start[0])
        if sx < 0 or sx >= columns:
            _fail("INVALID_INPUT")
        sy = _integral(start[1])
        if sy < 0 or sy >= rows:
            _fail("INVALID_INPUT")
    max_steps = _bounded_int(config["maxSteps"], 0, _MAX_STEPS)
    max_cells = _bounded_int(config["maxCells"], 0, _MAX_CELLS)
    seed, state = _read_random(config["random"])
    potential = len(starts) * (max_steps + 1)
    if potential > max_cells:
        raise LatticeError("WORK_LIMIT_EXCEEDED")

    start_x = [int(_pair(item)[0]) for item in starts]
    start_y = [int(_pair(item)[1]) for item in starts]

    stream = _Xoshiro(seed, state)
    out_x = array("i", [])
    out_y = array("i", [])
    offsets = [0]
    reasons = []
    occupied = set()

    for path in range(len(starts)):
        current_x, current_y = start_x[path], start_y[path]
        if (current_x, current_y) in occupied:
            offsets.append(len(out_x))
            reasons.append("occupied-start")
            continue
        occupied.add((current_x, current_y))
        out_x.append(current_x)
        out_y.append(current_y)
        if max_steps == 0:
            offsets.append(len(out_x))
            reasons.append("step-limit")
            continue
        moved = 0
        reason = None
        while moved < max_steps:
            candidates = []
            if current_y > 0 and (current_x, current_y - 1) not in occupied:
                candidates.append((current_x, current_y - 1))
            if current_x + 1 < columns and (current_x + 1, current_y) not in occupied:
                candidates.append((current_x + 1, current_y))
            if current_y + 1 < rows and (current_x, current_y + 1) not in occupied:
                candidates.append((current_x, current_y + 1))
            if current_x > 0 and (current_x - 1, current_y) not in occupied:
                candidates.append((current_x - 1, current_y))
            if not candidates:
                reason = "blocked"
                break
            chosen = stream.choose(len(candidates))
            current_x, current_y = candidates[chosen]
            occupied.add((current_x, current_y))
            out_x.append(current_x)
            out_y.append(current_y)
            moved += 1
            if moved == max_steps:
                reason = "step-limit"
        offsets.append(len(out_x))
        reasons.append(reason)
    return _LatticePaths(out_x, out_y, offsets, reasons, stream.state())

File: packages/javascript/src/occupied-lattice-paths.js
const MAX_CELLS = 1073741823;
const MAX_STEPS = 1073741822;
const MAX_SAFE = 9007199254740991;
const UINT32_MAX = 4294967295;

/** Error with one of the stable path.occupied-lattice-paths-2d contract codes. */
export class LatticeError extends Error {
  constructor(code) { super(code); this.name = "LatticeError"; this.code = code; }
}

function fail(code) { throw new LatticeError(code); }
function finite(value) { return typeof value === "number" && Number.isFinite(value); }
function integral(value) {
  if (!finite(value) || value !== Math.floor(value)) fail("INVALID_INPUT");
  return value;
}
function boundedInt(value, lo, hi) {
  const n = integral(value);
  if (n < lo || n > hi) fail("INVALID_INPUT");
  return n;
}
function uint32(value) { return boundedInt(value, 0, UINT32_MAX); }

function passiveRecord(value, keys) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail("INVALID_INPUT");
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) fail("INVALID_INPUT");
  if (Reflect.ownKeys(value).length !== keys.length) fail("INVALID_INPUT");
  for (const key of keys) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !("value" in descriptor)) fail("INVALID_INPUT");
  }
  return value;
}
function recordValue(record, key) { return Object.getOwnPropertyDescriptor(record, key).value; }
function hasOwnKey(value, key) {
  const descriptor = Object.getOwnPropertyDescriptor(value, key);
  return descriptor !== undefined && "value" in descriptor;
}
function passiveArray(value) {
  if (!Array.isArray(value) || Object.getPrototypeOf(value) !== Array.prototype) fail("INVALID_INPUT");
  return value;
}
function writableSlot(array, index) {
  const key = String(index);
  const descriptor = Object.getOwnPropertyDescriptor(array, key);
  if (descriptor !== undefined) return "value" in descriptor && descriptor.writable === true;
  if (!Object.isExtensible(array)) return false;
  for (let prototype = Object.getPrototypeOf(array); prototype !== null; prototype = Object.getPrototypeOf(prototype)) {
    if (Object.getOwnPropertyDescriptor(prototype, key) !== undefined) return false;
  }
  return true;
}
function arrayItem(array, index) {
  const descriptor = Object.getOwnPropertyDescriptor(array, String(index));
  if (descriptor === undefined || !("value" in descriptor)) fail("INVALID_INPUT");
  return descriptor.value;
}
function pair(value) {
  const array = passiveArray(value);
  if (array.length !== 2) fail("INVALID_INPUT");
  return array;
}
function recordEitherKey(value) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail("INVALID_INPUT");
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) fail("INVALID_INPUT");
  if (Reflect.ownKeys(value).length !== 1) fail("INVALID_INPUT");
  return value;
}

// Private xoshiro128**1.1 stream with two SplitMix64 expansion words; same
// algorithm as quadrant-partition.js/noise-band-path.js's Stream/Xoshiro, extended
// to accept an explicit caller-supplied [s0,s1,s2,s3] state that bypasses expansion.
class Xoshiro {
  constructor(seed, state) {
    if (state !== null) {
      this.s0 = state[0] >>> 0; this.s1 = state[1] >>> 0; this.s2 = state[2] >>> 0; this.s3 = state[3] >>> 0;
      return;
    }
    let x = BigInt(seed) + 0x9e3779b97f4a7c15n;
    const a = this.split(x);
    x += 0x9e3779b97f4a7c15n;
    const b = this.split(x);
    this.s0 = Number(a & 0xffffffffn) >>> 0;
    this.s1 = Number(a >> 32n) >>> 0;
    this.s2 = Number(b & 0xffffffffn) >>> 0;
    this.s3 = Number(b >> 32n) >>> 0;
  }
  split(x) {
    x = BigInt.asUintN(64, x);
    x = BigInt.asUintN(64, (x ^ (x >> 30n)) * 0xbf58476d1ce4e5b9n);
    x = BigInt.asUintN(64, (x ^ (x >> 27n)) * 0x94d049bb133111ebn);
    return BigInt.asUintN(64, x ^ (x >> 31n));
  }
  nextU32() {
    const value = Math.imul(this.s1, 5) >>> 0;
    const result = Math.imul(((value << 7) | (value >>> 25)) >>> 0, 9) >>> 0;
    const t = (this.s1 << 9) >>> 0;
    this.s2 = (this.s2 ^ this.s0) >>> 0;
    this.s3 = (this.s3 ^ this.s1) >>> 0;
    this.s1 = (this.s1 ^ this.s2) >>> 0;
    this.s0 = (this.s0 ^ this.s3) >>> 0;
    this.s2 = (this.s2 ^ t) >>> 0;
    this.s3 = ((this.s3 << 11) | (this.s3 >>> 21)) >>> 0;
    return result;
  }
  /** floor(u32 * count / 2^32); consumes exactly one word for count in 1..4. */
  choose(count) { return Math.floor((this.nextU32() * count) / 4294967296); }
  state() { return [this.s0, this.s1, this.s2, this.s3]; }
}

function readRandom(value) {
  const record = recordEitherKey(value);
  const hasSeed = hasOwnKey(record, "seed"), hasState = hasOwnKey(record, "state");
  if (hasSeed === hasState) fail("INVALID_INPUT");
  if (hasSeed) return { seed: uint32(recordValue(record, "seed")), state: null };
  const words = passiveArray(recordValue(record, "state"));
  if (words.length !== 4) fail("INVALID_INPUT");
  const state = [uint32(arrayItem(words, 0)), uint32(arrayItem(words, 1)), uint32(arrayItem(words, 2)), uint32(arrayItem(words, 3))];
  if (state[0] === 0 && state[1] === 0 && state[2] === 0 && state[3] === 0) fail("INVALID_INPUT");
  return { seed: 0, state };
}

class LatticePaths {
  #x; #y; #offsets; #reasons; #randomState;
  constructor(x, y, offsets, reasons, randomState) {
    this.#x = x; this.#y = y; this.#offsets = offsets; this.#reasons = reasons; this.#randomState = randomState;
    Object.freeze(this);
  }
  get pathCount() { return this.#reasons.length; }

  #pathIndex(path) {
    if (typeof path !== "number" || !Number.isFinite(path) || path < 0
        || path !== Math.floor(path) || path > MAX_SAFE) fail("INVALID_INDEX");
    if (path >= this.#reasons.length) fail("INDEX_OUT_OF_RANGE");
    return path;
  }
  #cellIndex(path, cell) {
    if (typeof cell !== "number" || !Number.isFinite(cell) || cell < 0
        || cell !== Math.floor(cell) || cell > MAX_SAFE) fail("INVALID_INDEX");
    const length = this.#offsets[path + 1] - this.#offsets[path];
    if (cell >= length) fail("INDEX_OUT_OF_RANGE");
    return cell;
  }
  pathLengthAt(path) {
    const p = this.#pathIndex(path);
    return this.#offsets[p + 1] - this.#offsets[p];
  }
  cellAt(path, cell) {
    const p = this.#pathIndex(path);
    const c = this.#cellIndex(p, cell);
    const at = this.#offsets[p] + c;
    return [this.#x[at], this.#y[at]];
  }
  cellInto(path, cell, output, offset = 0) {
    const p = this.#pathIndex(path);
    const c = this.#cellIndex(p, cell);
    if (!Array.isArray(output) && !(output instanceof Int32Array)) fail("INVALID_OUTPUT");
    if (!Number.isInteger(offset) || offset < 0 || offset > output.length - 2) fail("INVALID_OUTPUT");
    if (!writableSlot(output, offset) || !writableSlot(output, offset + 1)) fail("INVALID_OUTPUT");
    const at = this.#offsets[p] + c;
    output[offset] = this.#x[at];
    output[offset + 1] = this.#y[at];
    return output;
  }
  completionReasonAt(path) { return this.#reasons[this.#pathIndex(path)]; }
  randomState() { return this.#randomState.slice(); }

  toValues() {
    const paths = [];
    for (let p = 0; p < this.#reasons.length; p += 1) {
      const path = [];
      for (let i = this.#offsets[p]; i < this.#offsets[p + 1]; i += 1) path.push([this.#x[i], this.#y[i]]);
      paths.push(path);
    }
    return { paths, completionReasons: this.#reasons.slice(), randomState: this.#randomState.slice() };
  }
}

/**
 * Generate ordered retained cardinal cell paths whose cells are claimed by one
 * call-local occupancy set. Implements path.occupied-lattice-paths-2d 0.1.0
 * independently of source code. Motivating sketch: survey/out/2019/generativos/tata;
 * no artistic dimensions/starts/step range is established, see the catalog contract
 * for evidence.
 */
export function occupiedLatticePaths2D(config) {
  const record = passiveRecord(config, ["dimensions", "starts", "maxSteps", "maxCells", "random"]);
  const dimensions = pair(recordValue(record, "dimensions"));
  const columns = boundedInt(arrayItem(dimensions, 0), 1, 2147483647);
  const rows = boundedInt(arrayItem(dimensions, 1), 1, 2147483647);
  const starts = passiveArray(recordValue(record, "starts"));
  if (starts.length > MAX_CELLS) fail("INVALID_INPUT");
  for (let i = 0; i < starts.length; i += 1) {
    const start = pair(arrayItem(starts, i));
    const sx = integral(arrayItem(start, 0));
    if (sx < 0 || sx >= columns) fail("INVALID_INPUT");
    const sy = integral(arrayItem(start, 1));
    if (sy < 0 || sy >= rows) fail("INVALID_INPUT");
  }
  const maxSteps = boundedInt(recordValue(record, "maxSteps"), 0, MAX_STEPS);
  const maxCells = boundedInt(recordValue(record, "maxCells"), 0, MAX_CELLS);
  const random = readRandom(recordValue(record, "random"));
  const potential = BigInt(starts.length) * (BigInt(maxSteps) + 1n);
  if (potential > BigInt(maxCells)) throw new LatticeError("WORK_LIMIT_EXCEEDED");

  const startX = new Array(starts.length), startY = new Array(starts.length);
  for (let i = 0; i < starts.length; i += 1) {
    const start = pair(arrayItem(starts, i));
    startX[i] = arrayItem(start, 0);
    startY[i] = arrayItem(start, 1);
  }

  const stream = new Xoshiro(random.seed, random.state);
  const outX = [], outY = [], offsets = [0], reasons = [];
  const occupied = new Set();
  const key = (x, y) => x + "," + y;

  for (let path = 0; path < starts.length; path += 1) {
    let currentX = startX[path], currentY = startY[path];
    if (occupied.has(key(currentX, currentY))) {
      offsets.push(outX.length);
      reasons.push("occupied-start");
      continue;
    }
    occupied.add(key(currentX, currentY));
    outX.push(currentX); outY.push(currentY);
    if (maxSteps === 0) {
      offsets.push(outX.length);
      reasons.push("step-limit");
      continue;
    }
    let moved = 0;
    let reason = null;
    while (moved < maxSteps) {
      const candidates = [];
      if (currentY > 0 && !occupied.has(key(currentX, currentY - 1))) candidates.push([currentX, currentY - 1]);
      if (currentX + 1 < columns && !occupied.has(key(currentX + 1, currentY))) candidates.push([currentX + 1, currentY]);
      if (currentY + 1 < rows && !occupied.has(key(currentX, currentY + 1))) candidates.push([currentX, currentY + 1]);
      if (currentX > 0 && !occupied.has(key(currentX - 1, currentY))) candidates.push([currentX - 1, currentY]);
      if (candidates.length === 0) { reason = "blocked"; break; }
      const chosen = stream.choose(candidates.length);
      [currentX, currentY] = candidates[chosen];
      occupied.add(key(currentX, currentY));
      outX.push(currentX); outY.push(currentY);
      moved += 1;
      if (moved === maxSteps) reason = "step-limit";
    }
    offsets.push(outX.length);
    reasons.push(reason);
  }
  return new LatticePaths(outX, outY, offsets, reasons, stream.state());
}
